File: ios/Classes/CDAdPlatformView.swift
import Chalkdigital_Mobile_Ads_SDK
import Flutter
import Foundation

class CDAdPlatformView: NSObject, FlutterPlatformView, CDAdViewDelegate {
    private var containerView: UIView
    private var adView: CDAdView
    private var channel: FlutterMethodChannel
    private var placementId: String?
    private var partnerId: String?
    private var isLocationAutoUpdateEnabled: Bool = true
    private var isAdAutoRefreshEnabled: Bool = true
    private var isTesting: Bool = false
    private var isCloseable: Bool = true
    private var useInAppBrowser: Bool = true
    private var cdGeoInfo: CDGeoInfo?

    init(
        frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?,
        messenger: FlutterBinaryMessenger
    ) {

        self.containerView = UIView(
            frame: CGRect(x: 0, y: 0, width: frame.width + 32, height: frame.height + 16))
        self.containerView.backgroundColor = .clear
        self.adView = CDAdView(frame: frame)
        self.containerView.addSubview(adView)
        self.adView.translatesAutoresizingMaskIntoConstraints = false

        NSLayoutConstraint.activate([
            self.adView.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 16),
            self.adView.leadingAnchor.constraint(
                equalTo: containerView.leadingAnchor, constant: 16),
            self.adView.trailingAnchor.constraint(
                equalTo: containerView.trailingAnchor, constant: -16),
        ])
        self.channel = FlutterMethodChannel(
            name: "cdads_ad_view_\(viewId)", binaryMessenger: messenger)
        super.init()

        channel.setMethodCallHandler { [weak self] call, result in
            guard let self = self else { return }

            switch call.method {
            case "loadAd":
                if let args = call.arguments as? [String: Any] {
                    if let placementId = args["placementId"] as? String {
                        self.placementId = placementId
                    }

                    if let partnerId = args["partnerId"] as? String {
                        self.partnerId = partnerId
                    }

                    if let locationEnabled = args["isLocationAutoUpdateEnabled"] as? Bool {
                        self.isLocationAutoUpdateEnabled = locationEnabled
                    }

                    if let refreshEnabled = args["isAdAutoRefreshEnabled"] as? Bool {
                        self.isAdAutoRefreshEnabled = refreshEnabled
                    }

                    if let testing = args["isTesting"] as? Bool {
                        self.isTesting = testing
                    }
                    if let closeable = args["isCloseable"] as? Bool {
                        self.isCloseable = closeable
                    }
                    if let useInAppBrowser = args["useInAppBrowser"] as? Bool {
                        self.useInAppBrowser = useInAppBrowser
                    }

                    if let cdGeoMap = args["geoInfo"] as? [String: Any] {
                        let cdGeoInfo = CDGeoInfo()

                        cdGeoInfo.lat = (cdGeoMap["lat"] as? NSNumber)?.floatValue ?? 0.0
                        cdGeoInfo.lon = (cdGeoMap["lon"] as? NSNumber)?.floatValue ?? 0.0
                        cdGeoInfo.type = (cdGeoMap["lon"] as? NSNumber)?.integerValue ?? 0.0
                        cdGeoInfo.countryCode = cdGeoMap["countryCode"] as? String ?? ""
                        cdGeoInfo.region = cdGeoMap["region"] as? String ?? ""
                        cdGeoInfo.city = cdGeoMap["city"] as? String ?? ""
                        cdGeoInfo.zip = cdGeoMap["zip"] as? String ?? ""
                        cdGeoInfo.streetAddress = cdGeoMap["streetAddress"] as? String ?? ""
                        cdGeoInfo.subThoroughFare = cdGeoMap["subThoroughFare"] as? String ?? ""
                        cdGeoInfo.haccuracy = cdGeoMap["haccuracy"] as? String ?? ""
                    }
                }

                self.adView.getAdWithNotification()
                result(nil)

            default:
                result(FlutterMethodNotImplemented)
            }
        }

        self.adView.delegate = self
        self.adView.getAdWithNotification()  // Start ad load
    }

    func view() -> UIView {
        return adView
    }

    // MARK: - CDAdViewDelegate required methods
    func placementId(_ cdAdView: CDAdView!) -> String! {
        return self.placementId ?? "0"
    }

    func partnerId(_ cdAdView: CDAdView!) -> String! {
        return self.partnerId ?? ""
    }

    func locationServicesEnabled(_ cdAdView: CDAdView!) -> Bool {
        return self.isLocationAutoUpdateEnabled
    }
    func allowAutomaticRefreshAds(_ cdAdView: CDAdView!) -> Bool {
        return self.isAdAutoRefreshEnabled
    }

    func closeable(_ cdAdView: CDAdView!) -> Bool {
        return self.isCloseable
    }

    func use(inAppBrowser cdAdView: CDAdView!) -> Bool {
        return self.useInAppBrowser
    }

    func cdAdViewFrame(_ cdAdView: CDAdView!) -> CGRect {
        print(adView.frame.debugDescription)
        return adView.frame
    }

    func cdAdViewSize(_ cdAdView: CDAdView!) -> CDAdSize {
        return CDAdSize(size: adView.frame.size)
    }

    func locationServicesEnabled(_ cdAdView: CDAdView!) -> Bool {
        return true
    }

    func use(inAppBrowser cdAdView: CDAdView!) -> Bool {
        return true
    }

    func applicationUIViewController(_ cdAdView: CDAdView!) -> UIViewController! {
        // Use root view controller
        return UIApplication.shared.delegate?.window??.rootViewController
    }

    // MARK: - Optional delegate callbacks sent to Flutter
    func getAdSucceeded(_ cdAdView: CDAdView!) {
        channel.invokeMethod("onAdLoaded", arguments: nil)
    }

    func getAdFailed(_ cdAdView: CDAdView!) {
        channel.invokeMethod("onAdFailed", arguments: nil)
    }

    func adViewDidReceiveTapEvent(_ cdAdView: CDAdView!) {
        channel.invokeMethod("onAdTapped", arguments: nil)
    }

    func interstitialClosed(_ cdAdView: CDAdView!) {
        channel.invokeMethod("onInterstitialClosed", arguments: nil)
    }

    // Add other callbacks as needed...
}

File: android/src/main/kotlin/com/example/cdads/CDAdPlatformView.kt
import android.content.Context
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import com.chalkdigital.ads.CDAdErrorCode
import com.chalkdigital.ads.CDAdView
import com.chalkdigital.ads.CDAdView.CDAdViewListener
import com.chalkdigital.common.CDAdGeoInfo
import com.chalkdigital.common.CDAdSize
import com.chalkdigital.common.CDAdsUtils
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView


class CDAdPlatformView(
    private val context: Context,
    viewId: Int,
    creationParams: Map<String, Any>,
    messenger: io.flutter.plugin.common.BinaryMessenger
) : PlatformView, CDAdViewListener {

    private val cdAdView: CDAdView = CDAdView(context)
    private val containerView: LinearLayout = LinearLayout(context);

    private val channel: MethodChannel = MethodChannel(messenger, "cdads_ad_view_$viewId")

    private var placementId: String = "0"
    private var partnerId: String = ""
    private var isLocationAutoUpdateEnabled: Boolean = true
    private var isAdAutoRefreshEnabled: Boolean = true
    private var isTesting: Boolean = false
    private var isCloseable: Boolean = true
    private var useInAppBrowser: Boolean = true

    init {
        containerView.setOrientation(LinearLayout.VERTICAL);
        // Set layout parameters or size if needed
        containerView.layoutParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )

        val layoutParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        cdAdView.layoutParams = layoutParams
        containerView.addView(cdAdView)
        cdAdView.cdAdViewListener = this
        @Suppress("UNCHECKED_CAST") val adSize : Map<String, Any> = creationParams["adSize"] as Map<String, Any>
        cdAdView.cdAdSize = CDAdSize.getCDAdSizeConstantFromSize(adSize["width"] as Int, adSize["height"] as Int);
        CDAdsUtils.setBrowserAgent(
            if (useInAppBrowser) CDAdsUtils.BrowserAgent.IN_APP else CDAdsUtils.BrowserAgent.NATIVE
        )

        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "loadAd" -> {
                    val args = call.arguments as? Map<*, *>
                    args?.let { map ->
                        placementId = map["placementId"] as? String ?: placementId
                        partnerId = map["partnerId"] as? String ?: partnerId
                        isLocationAutoUpdateEnabled = map["isLocationAutoUpdateEnabled"] as? Boolean ?: isLocationAutoUpdateEnabled
                        isAdAutoRefreshEnabled = map["isAdAutoRefreshEnabled"] as? Boolean ?: isAdAutoRefreshEnabled
                        isTesting = map["isTesting"] as? Boolean ?: isTesting
                        isCloseable = map["isCloseable"] as? Boolean ?: isCloseable
                        useInAppBrowser = map["useInAppBrowser"] as? Boolean ?: useInAppBrowser
                        val cdGeoInfo = CDAdGeoInfo()
                        val cdGeoMap = call.argument("geoInfo") as Map<String, Any>?
                        if(cdGeoMap != null && cdGeoMap.size>0){
                            cdGeoInfo.lat = (cdGeoMap["lat"] as Double).toFloat()
                            cdGeoInfo.lon = (cdGeoMap["lon"] as Double).toFloat()
                            cdGeoInfo.type = (cdGeoMap["type"] as Int).toInt()
                            cdGeoInfo.countryCode = cdGeoMap["countryCode"] as String?
                            cdGeoInfo.region = cdGeoMap["region"] as String?
                            cdGeoInfo.city = cdGeoMap["city"] as String?
                            cdGeoInfo.zip = cdGeoMap["zip"] as String?
                            cdGeoInfo.accuracy = cdGeoMap["haccuracy"] as String?
                            cdGeoInfo.time = System.currentTimeMillis()
                            cdAdView.setCDAdGeoInfo(cdGeoInfo)
                        }
                        cdAdView.isLocationAutoUpdateEnabled = isLocationAutoUpdateEnabled
                        cdAdView.setAdAutoRefreshEnabled(isAdAutoRefreshEnabled)
                        cdAdView.setTesting(isTesting)

                    }
                    // Trigger ad load
                    cdAdView.requestNewAd(HashMap(), partnerId, placementId)
                    result.success(null)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }
//        cdAdView.requestNewAd(HashMap(),"chalkboard", "" )

    }

    fun dpToPx(context: Context, dp: Int): Int {
        return (dp * context.resources.displayMetrics.density).toInt()
    }

    override fun getView(): View = containerView

    override fun dispose() {
        cdAdView.destroy() // If your SDK has a destroy method
    }

    override fun onBannerAdRequest(banner: CDAdView?) {

    }

    override fun onBannerLoaded(banner: CDAdView?) {
        channel.invokeMethod("onAdLoaded", null)
    }

    override fun onBannerFailed(banner: CDAdView?, errorCode: CDAdErrorCode?) {
//        val errorMap = mapOf("errorCode" to (errorCode?.name ?: "UNKNOWN"))
        channel.invokeMethod("onAdFailed", null)
    }

    override fun onBannerClicked(banner: CDAdView?) {
        channel.invokeMethod("onAdTapped", null)
    }

    override fun onBannerExpanded(banner: CDAdView?) {

    }

    override fun onBannerCollapsed(banner: CDAdView?) {

    }
}
